<a href="https://colab.research.google.com/github/Saifuddin-Yasir/ML-AI-Learning-Journey/blob/main/DL_Assignment_03_Question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# DL Assignment 03

**Name: Saifuddin Yasir**

**Course Email: saifuddinnyasir20@gmail.com**  


## End of Assignment

Before submitting:
- Run all cells from top to bottom.  
- Check that all answer sections are filled.  
- Instruction video অনুযায়ী আমাদের দেয়া Colab ফাইলটি থেকে প্রথম একটি Save copy in drive করে নিবা। এরপর Google colab এর মধ্যে কোডগুলো করবে এবং সেই ফাইলটি ‘Anyone with the link’ & ‘View’ Access দিয়ে ফাইলটির Shareble Link টি সাবমিট করবে।

# General Instruction

You must choose your own dataset.

The dataset must:

Be a supervised learning dataset (Regression or Binary Classification)

Contain at least 300 samples

Have at least 2 input features

Be in CSV format

You are NOT allowed to use Dataset or DataLoader.

You must implement everything manually.

# Question 01: [ Marks 05 ]

## Dataset Preparation

## Using your chosen dataset:

Load the dataset.

Perform necessary preprocessing:

Handle missing values (if any)

Encode categorical variables (if necessary)

Feature scaling (if needed)

Separate features (X) and target (y).

Convert them into NumPy arrays.

Convert them into PyTorch tensors.

Split into training and testing sets.

Clearly explain each preprocessing decision.

# **Write** Answer 01:


In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
import torch

In [ ]:
data= pd.read_csv('/content/water_potability.csv')
data.head()

,ph,Hardness,Solids,Chloramines,Sulfate,Conductivity,Organic_carbon,Trihalomethanes,Turbidity,Potability
0,NaN,204.890455,20791.318981,7.300212,368.516441,564.308654,10.379783,86.990970,2.963135,0
1,3.716080,129.422921,18630.057858,6.635246,NaN,592.885359,15.180013,56.329076,4.500656,0
2,8.099124,224.236259,19909.541732,9.275884,NaN,418.606213,16.868637,66.420093,3.055934,0
3,8.316766,214.373394,22018.417441,8.059332,356.886136,363.266516,18.436524,100.341674,4.628771,0
4,9.092223,181.101509,17978.986339,6.546600,310.135738,398.410813,11.558279,31.997993,4.075075,0


In [ ]:
data['ph']=data['ph'].fillna(data['ph'].mean())
data['Sulfate']=data['Sulfate'].fillna(data['Sulfate'].mean())
data['Trihalomethanes']=data['Trihalomethanes'].fillna(data['Trihalomethanes'].mean())

In [ ]:
numerical = ['ph', 'Hardness', 'Solids', 'Chloramines', 'Sulfate', 'Conductivity', 'Organic_carbon','Trihalomethanes', 'Turbidity']
scaling = StandardScaler()
scaling.fit(data[numerical])
data[numerical] = scaling.transform(data[numerical])

In [ ]:
x = data.drop('Potability', axis=1)
y = data['Potability']

In [ ]:
x = x.to_numpy()
y = y.to_numpy()

In [ ]:
x_train,x_test, y_train, y_test = train_test_split(x,y,test_size=0.2,random_state=42)

In [ ]:
x_train= torch.tensor(x_train, dtype=torch.float32)
x_test= torch.tensor(x_test, dtype=torch.float32)
y_train= torch.tensor(y_train, dtype=torch.float32)
y_test= torch.tensor(y_test, dtype=torch.float32)

In [ ]:
x_train.shape[1]

9

####Explain Preprocessing steps

After loading dataset using pandas, first preprocessing step applied missing data handling on three input feature (ph, Sulfate, Trihalomethanes). then,feature scaling was applied using standard scaling on all input features for priotirize all features equally.It's use when there is a large number differences between input features. there is no categorical features in this dataset. So, No categorical encoding required. Then, Separate input features and target features. first convert them to numpy array and then, convert to tensor array for using pytorch library. lastly, we split the test and train data and 20/80 ratio.

# Question 02: [ Marks 20 ]

## Design a neural network using nn.Module.

### The model must contain:

Input layer

At least one hidden layer

Output layer

Suitable activation function



## Justify:

Number of hidden neurons

Choice of activation function

Print  the total number of trainable parameters.


## Write Answer 02:


In [ ]:
import torch.nn as nn

In [ ]:
class Model(nn.Module):
  def __init__(self,numerical):
   super().__init__()

   self.network= nn.Sequential(
   nn.Linear(numerical,27) ,
   nn.ReLU(),
   nn.Linear(27,18) ,
   nn.ReLU(),
   nn.Linear(18,9) ,
   nn.ReLU(),
   nn.Linear(9,1),
   nn.Sigmoid())

  def forward(self,x):
    out= self.network(x)

    return out

In [ ]:
model=Model(x_train.shape[1])

In [ ]:
model.forward(x_train)

tensor([[0.5438],
        [0.5381],
        [0.5366],
        ...,
        [0.5408],
        [0.5371],
        [0.5437]], grad_fn=<SigmoidBackward0>)

In [ ]:
from torchinfo import summary
summary(model,input_size=(300,9))

Layer (type:depth-idx)                   Output Shape              Param #
Model                                    [300, 1]                  --
├─Sequential: 1-1                        [300, 1]                  --
│    └─Linear: 2-1                       [300, 27]                 270
│    └─ReLU: 2-2                         [300, 27]                 --
│    └─Linear: 2-3                       [300, 18]                 504
│    └─ReLU: 2-4                         [300, 18]                 --
│    └─Linear: 2-5                       [300, 9]                  171
│    └─ReLU: 2-6                         [300, 9]                  --
│    └─Linear: 2-7                       [300, 1]                  10
│    └─Sigmoid: 2-8                      [300, 1]                  --
Total params: 955
Trainable params: 955
Non-trainable params: 0
Total mult-adds (Units.MEGABYTES): 0.29
Input size (MB): 0.01
Forward/backward pass size (MB): 0.13
Params size (MB): 0.00
Estimated Total Size (MB): 0.15

#### Justify

There are 3 hidden layer used in the neural network. 1st hidden layer have 27 neurons, 2nd one have 18 neurons and last hidden layer have 9 neurons. the more neurons in each hidden layer, it can learn more complex pattern from input features and predict with good accuracy.

 The dataset is a binary classification type where it predict only potable or not. so in output layer we use sigmoid as activation function which give the probility of water as outputs.and in hidden layer, i used ReLU which can learn non-linearity pattern very well from data.

 here, the total trainable parameter is 955.

# Question 03: [ Marks 10 ]

Choose an appropriate loss function.

Choose an optimizer.

<br>

Justify your choices based on:

Regression vs Classification

Nature of the dataset

## Write Answer 03:

In [ ]:
loss_function= nn.BCELoss()
optimizer= torch.optim.Adam(model.parameters(),lr=0.01)

The dataset is about binary classification type dataset. for this type of data, BCE is most commonly used as a loss function. it measure difference of true and predicted pobability. from it's equation, it's increased loss when the pobabiity and true label become far from each other.


Here, i use adam optimizer, because, it adapts learning rate for each parameter and also it's faster. it's genarally used as a default.

# Question 04: [ Marks 15 ]

## Implement a full training loop:

Forward pass

Loss computation

Backward pass

Parameter update

Gradient reset

### Requirements:

Train for at least 100 epochs.

Print loss every 10 epochs.

Store training loss history(You can pick your own Data Structure).

Explain clearly what happens in each step of the pipeline.

## Write Answer 04:

In [ ]:
epochs=500

for epoch in range(epochs):

  y_pred= model(x_train)
  loss= loss_function(y_pred,y_train.reshape(-1,1))
  optimizer.zero_grad()
  loss.backward()
  optimizer.step()

  if epoch % 10 == 0:
    print(f'Epoch: {epoch}, Loss: {loss.item()}')

Epoch: 0, Loss: 0.7128764390945435
Epoch: 10, Loss: 0.6661279201507568
Epoch: 20, Loss: 0.6331451535224915
Epoch: 30, Loss: 0.5918014645576477
Epoch: 40, Loss: 0.5749120116233826
Epoch: 50, Loss: 0.5619109272956848
Epoch: 60, Loss: 0.5467222332954407
Epoch: 70, Loss: 0.5304581522941589
Epoch: 80, Loss: 0.5143346786499023
Epoch: 90, Loss: 0.5009410381317139
Epoch: 100, Loss: 0.49038010835647583
Epoch: 110, Loss: 0.4835330545902252
Epoch: 120, Loss: 0.4762673079967499
Epoch: 130, Loss: 0.46969759464263916
Epoch: 140, Loss: 0.4671576917171478
Epoch: 150, Loss: 0.46035537123680115
Epoch: 160, Loss: 0.4556792080402374
Epoch: 170, Loss: 0.4511505961418152
Epoch: 180, Loss: 0.44703957438468933
Epoch: 190, Loss: 0.45085254311561584
Epoch: 200, Loss: 0.44226208329200745
Epoch: 210, Loss: 0.4364506006240845
Epoch: 220, Loss: 0.43557339906692505
Epoch: 230, Loss: 0.4332551658153534
Epoch: 240, Loss: 0.426349401473999
Epoch: 250, Loss: 0.4213811755180359
Epoch: 260, Loss: 0.4187214970588684
Epoch:

In each steps in the pipeline:



1.   forward propagration was done and got y_pred
2.   then, BCE is calculated by y_pred and y_train which stored in loss variable.

1.  After that, we clear gradient by using zero_grad() before doing back propagation.
2.  New gradient was calculate by using backward() function.

1.   the new weight and biases are updated by using optimizer function.
2.   lastly, we showed losses after each 10 epochs.







# Question 05: [ Marks 10 ]

## Evaluate the model on test data.

## For regression:

Report MSE and MAE


## For classification:

Report Accuracy

Compare training vs testing performance.

State whether the model is underfitting or overfitting.

## Write Answer 05:

In [ ]:
with torch.no_grad():
  y_pred= model.forward(x_test)
  y_pred= (y_pred >.5).float()
  accuracy_test= (y_pred==y_test).float().mean()
  print(f'test Accuracy: {accuracy_test}')

  y_pred= model.forward(x_train)
  y_pred= (y_pred >.5).float()
  accuracy_train= (y_pred==y_test).float().mean()
  print(f'train Accuracy: {accuracy_train}')

test Accuracy: 0.528498649597168
train Accuracy: 0.5334295034408569


As test accuracy and train accuracy is same, neither the model is overfitted nor underfitted.

# Question 06: [ Marks 20 ]

## Modify at least ONE of the following:

Learning rate

Number of hidden neurons

Number of epochs

### Train again and compare:

Convergence speed

Final performance

Explain how the change affected the model.

## Write Answer 06:

In [ ]:
class Model(nn.Module):
  def __init__(self,numerical):
   super().__init__()

   self.network= nn.Sequential(
   nn.Linear(numerical,18),
   nn.ReLU(),
   nn.Linear(18,9),
   nn.ReLU(),
   nn.Linear(9,9),
   nn.ReLU(),
   nn.Linear(9,1),
   nn.Sigmoid())

  def forward(self,x):
    out= self.network(x)
    return out

model=Model(x_train.shape[1])
model(x_train)

epochs=500

for epoch in range(epochs):

  y_pred= model(x_train)
  loss= loss_function(y_pred,y_train.reshape(-1,1))
  optimizer.zero_grad()
  loss.backward()
  optimizer.step()

  if epoch % 10 == 0:
    print(f'Epoch: {epoch}, Loss: {loss.item()}')

with torch.no_grad():
  y_pred= model.forward(x_test)
  y_pred= (y_pred >0.5).float()
  accuracy= (y_pred==y_test.reshape(1,-1)).float().mean()
  print(f'Accuracy: {accuracy}')

Epoch: 0, Loss: 0.6963135600090027
Epoch: 10, Loss: 0.6963135600090027
Epoch: 20, Loss: 0.6963135600090027
Epoch: 30, Loss: 0.6963135600090027
Epoch: 40, Loss: 0.6963135600090027
Epoch: 50, Loss: 0.6963135600090027
Epoch: 60, Loss: 0.6963135600090027
Epoch: 70, Loss: 0.6963135600090027
Epoch: 80, Loss: 0.6963135600090027
Epoch: 90, Loss: 0.6963135600090027
Epoch: 100, Loss: 0.6963135600090027
Epoch: 110, Loss: 0.6963135600090027
Epoch: 120, Loss: 0.6963135600090027
Epoch: 130, Loss: 0.6963135600090027
Epoch: 140, Loss: 0.6963135600090027
Epoch: 150, Loss: 0.6963135600090027
Epoch: 160, Loss: 0.6963135600090027
Epoch: 170, Loss: 0.6963135600090027
Epoch: 180, Loss: 0.6963135600090027
Epoch: 190, Loss: 0.6963135600090027
Epoch: 200, Loss: 0.6963135600090027
Epoch: 210, Loss: 0.6963135600090027
Epoch: 220, Loss: 0.6963135600090027
Epoch: 230, Loss: 0.6963135600090027
Epoch: 240, Loss: 0.6963135600090027
Epoch: 250, Loss: 0.6963135600090027
Epoch: 260, Loss: 0.6963135600090027
Epoch: 270, 

Here, I lower the number of neuron in hidden layers. Then, after traing the accuracy of model lower from 0.52 to 0.41. before the loss was 0.37 which incresed after lower neuron n hidden layer.Now, it's 0.69. As i didn't change the learning rate of model, the convergence speed was same before and after. the final performance become lower than before. Overall, lower number of neuron lower the accuracy and efficiency of model.

# Question 07: [ Marks 20 ]


# Training Analysis

Answer the following:

Why must gradients be reset every epoch?

What happens if learning rate is too high?

What happens if learning rate is too small?

Why do we define layers inside the constructor (__init__) and not inside forward()?


## Write Answer 07:



*   In pytorch backward() function the gradient is added with old gradient which give wrong result for a model if we don't reset before. So, we need to reset gradient before backward() function.

*  if the learning rate is too high, it takes larger step to reach global minima.So, it might cause oscillation also can miss the global minima. Also it might overshoot from global minima.


*   if learning rate is too small, it takes smaller step to converge and reach to global minima. It need too much time to reach global minima. sometimes,it got stuck on local minima and also It also seems stop on middle for very slow convergence to reach global minima for very small change. Thus, it fail to converge.

* the layer define of constructor(init) under a class, for created and tracked parameter. This things can't be tracked in forward function. it's also allow optimizer to update parameter. In function like forward(),it isn't possible to track or update. it's just create structure of input to output flows in network.



